File: simple-gemm.jl/src/gemm-dense-threads.jl

import Random
import Dates
import BenchmarkTools
import InteractiveUtils

@doc """
Generate random values from global random number generator
"""
function fill_random(A::Array{Float32,2})
    return Random.rand!(A)
end

@doc """
Simplified gemm: C = alpha A x B + C where alpha = 1 , C = zeros, 
so:
C = A x B 
"""
function gemm!(A::Array{Float32,2}, B::Array{Float32,2}, C::Array{Float32,2})
    A_rows = size(A)[2]
    A_cols = size(A)[1]
    B_rows = size(B)[2]


    Base.Threads.@threads for i = 1:A_rows
        for k = 1:A_cols
            for j = 1:B_rows
                C[j, i] += A[k, i] * B[j, k]
            end
        end
    end

    return
end

function main(args::Array{String,1})

    # must initialize scalars
    A_rows::Int32 = -1
    A_cols::Int32 = -1
    B_rows::Int32 = -1
    B_cols::Int32 = -1

    @show args

    # args don't include Julia executable and program
    if size(args)[1] != 3
        throw(
            ArgumentError(
                "Usage: 3 arguments: matrix A rows, matrix A cols and matrix B cols",
            ),
        )
    else
        A_rows = parse(Int32, args[1])
        A_cols = parse(Int32, args[2])
        B_rows = parse(Int32, args[2])
        B_cols = parse(Int32, args[3])
    end

    # Julia is column-based (like Fortran)
    A = Array{Float32,2}(undef, A_cols, A_rows)
    B = Array{Float32,2}(undef, B_cols, B_rows)
    C = zeros(Float32, B_cols, A_rows)

    fill_random(A)
    fill_random(B)

    BenchmarkTools.@time gemm!(A, B, C)
    #InteractiveUtils.@code_llvm gemm!(A, B, C)
    #InteractiveUtils.@code_native gemm!(A, B, C)


    # println(A)
    # println(B)
    # println(C)

end

main(ARGS)

File: python/GemmDenseThreads/GemmDenseThreads.py

import sys
import typing
from numba import njit, prange
import numpy as np
import time


@njit(parallel=True, nogil=True, fastmath=True)
def gemm(A: np.ndarray, B: np.ndarray, C: np.ndarray):

    A_rows = A.shape[0]
    A_cols = A.shape[1]
    B_cols = B.shape[1]

    # here set the parallel range "prange"
    for i in prange(0, A_rows):
    #for i in range(0, A_rows):
        for k in range(0, A_cols):
            temp = A[i, k]
            for j in range(0, B_cols):
                C[i, j] += temp * B[k, j]

    return

def _print_time(start, process:str):
    end = time.time()
    print("Time to " + process + " : " + str(end-start) + " s")
    return end 

def main():

    # must initialize scalars
    A_rows: int = -1
    A_cols: int = -1
    B_rows: int = -1
    B_cols: int = -1

    args = sys.argv[1:]
    print(args)

    # args don't include Julia executable and program
    if len(args) != 3:
        raise ValueError(
            "Usage: 3 arguments: matrix A rows, matrix A cols and matrix B cols")
    else:
        A_rows = int(args[0])
        A_cols = int(args[1])
        B_rows = int(args[1])
        B_cols = int(args[2])

    rng = np.random.default_rng()
    start = time.time()
    A = rng.random((A_rows, A_cols), dtype=np.float32)
    tmp = _print_time(start, "initialize A")
    B = rng.random((B_rows, B_cols), dtype=np.float32)
    tmp = _print_time(tmp, "initialize B")
    C = np.zeros(dtype=np.float32, shape=(A_rows, B_cols))
    tmp = _print_time(tmp, "initialize C")

    gemm(A, B, C)
    tmp = _print_time(tmp, "simple gemm")
    tmp = _print_time(start, "total time")
    

    # print(C)

    return 0


if __name__ == "__main__":
    main()
